src/rust/engine/graph/src/entry.rs

```diff
@@ -377,98 +384,135 @@ impl<N: Node> Entry<N> {
   /// need to consume the state (which avoids cloning some of the values held there), so we take it
   /// by value.
   ///
-  #[allow(clippy::type_complexity)] // This return type is not particularly complex.
   pub(crate) fn get_node_result(
     &mut self,
     context: &N::Context,
     entry_id: EntryId,
-  ) -> BoxFuture<Result<(N::Item, Generation), N::Error>> {
-    {
-      let mut state = self.state.lock();
+  ) -> BoxFuture<NodeResult<N>> {
+    let mut state = self.state.lock();
 
-      // First check whether the Node is already complete, or is currently running: in both of these
-      // cases we return early without swapping the state of the Node.
-      match *state {
-        EntryState::Running {
-          ref mut waiters, ..
-        } => {
-          let (send, recv) = oneshot::channel();
-          waiters.push(send);
-          return async move { recv.await.map_err(|_| N::Error::invalidated())? }.boxed();
-        }
-        EntryState::Completed {
-          ref result,
-          generation,
-          ..
-        } if result.is_clean(context) => {
-          return future::ready(Ok((result.as_ref().clone(), generation))).boxed();
+    // First check whether the Node is already complete, or is currently running: in both of these
+    // cases we return early without swapping the state of the Node.
+    match *state {
+      EntryState::Running {
+        ref pending_value, ..
+      } => {
+        if let Some(receiver) = pending_value.receiver() {
+          return async move { receiver.recv().await.ok_or_else(N::Error::invalidated)? }.boxed();
         }
-        _ => (),
-      };
+        // Else: this node was just canceled: fall through to restart it.
+      }
+      EntryState::Completed {
+        ref result,
+        generation,
+        ..
+      } if result.is_clean(context) => {
+        return future::ready(Ok((result.as_ref().clone(), generation))).boxed();
+      }
+      _ => (),
+    };
 
-      // Otherwise, we'll need to swap the state of the Node, so take it by value.
-      let next_state = match mem::replace(&mut *state, EntryState::initial()) {
-        EntryState::NotStarted {
-          run_token,
-          generation,
-          previous_result,
-        } => Self::spawn_node_execution(
+    // Otherwise, we'll need to swap the state of the Node, so take it by value.
+    let (next_state, receiver) = match mem::replace(&mut *state, EntryState::initial()) {
+      EntryState::NotStarted {
+        run_token,
+        generation,
+        previous_result,
+      }
+      | EntryState::Running {
```

Comment:
Checking my understanding: This is the fallthrough mentioned above - Running but cancelled (the pending value was dropped); thus, spawning a new execution is in fact the desired result.

Reply:
Correct!